File: CivitasRuby/lib/sorpresa_ir_a_casilla.rb
# encoding:utf-8s
require_relative 'sorpresa'
module Civitas
  class SorpresaIrACasilla < Sorpresa
    def initialize(texto,tablero,valor)
      super(texto)
      @tablero = tablero
      @valor = valor
    end
    
    
    def informe(actual, todos)
      Diario.instance.ocurre_evento("Aplicando sorpresa " + self.to_s + 
        " al jugador " + todos[actual].nombre + " de tipo IrACasilla")
    end
    
    
    def aplicar_a_jugador(actual, todos)
      casilla_actual = todos[actual].num_casilla_actual
      if(super.jugador_correcto(actual,todos))
        informe(actual,todos)
        tirada = @tablero.calcular_tirada(casilla_actual,@valor)
        nueva_posicion = @tablero.nueva_posicion(casilla_actual,tirada)
        todos[actual].mover_a_casilla(nueva_posicion)
        @tablero.casillas[nueva_posicion].recibe_jugador(actual,todos)
      end
    end
    
    def to_s
      "Sorpresa: { \n Texto: #{@texto}  \n Tipo: IrACasilla   \n Valor: #{@valor} }"
    end
  end
end

File: CivitasJava/src/Civitas/SorpresaIrACasilla.java
package Civitas;

import java.util.ArrayList;

public class SorpresaIrACasilla extends Sorpresa{
    private String texto;
    private int valor;
    private Tablero tablero;
    
    SorpresaIrACasilla(Tablero tablero, int valor, String texto){     
        super(texto);
        this.tablero = tablero;
        this.valor = valor;  
    }
    
    @Override
    void aplicarAJugador (int actual, ArrayList<Jugador> todos){
        int casillaActualNum = todos.get(actual).getNumCasillaActual();
        if(super.jugadorCorrecto(actual,todos)){
            informe(actual,todos);
            int tirada = tablero.calcularTirada(casillaActualNum, this.valor);
            int nuevaPosicion = tablero.nuevaPosicion(casillaActualNum, tirada);
            todos.get(actual).moverACasilla(nuevaPosicion);
            tablero.getCasilla(nuevaPosicion).recibeJugador(actual,todos);
        }
    }
    
    @Override
    public String toString(){
        return "Sorpresa: { \n Texto = " + texto + "\n Tipo = IrACasilla" + 
                "\n Valor = " + valor + "}";
    }
}
